### ISSUE TITLE
Implement Single Observable variant

### ISSUE THREAD

**Issue by mmacfadden** [Reactive-Extensions/RxJS#1349]
I was wondering if the RxJS community had considered an implementation of the Single variant of the Observable.  I realize that this is similar to a Promise.  But at the moment our API is left with two choices.

1) Have a mix of Promise and Observable.  This is not ideal since you can't really use them together, and promises are much less powerful.

2) Turn a promise into an Observable.  Also not idea because the API semantics of an observable (next, complete, error) are different than a promise / single (success/value, failure / error).  Simply returning an Observable instead of a promise muddies the API if the method really does always return exactly one value or an error.

I would love to eventually ditch the promise altogether  and use something like Single.

Thanks.

P.S. RxJS rules.

**Comment by corps** [comment]
> 1) Have a mix of Promise and Observable. This is not ideal since you can't really use them together, and promises are much less powerful.

You can use them together, but because they are really two separate concepts you have to cognitive about the semantics of both.  RxJS provides a Promise -> Observable utility method.  However, because Observables are more powerful than Promises there isn't always a simple mapping from Observable to Promises without making assumptions that are hard to generalize (Observables that complete with no values could map to a Promise that provides undefined?  Or it could map to a failure depending on your perspective).

Thus, my suggestion is to... just ditch Promises and stick to multi value Observables.  If you're only concerned about when something is complete, subscribe to complete.  If you're concerned about precisely the last value, you can use .take(1).  More often than not, however, data flows are not about "getting exactly one value" and often "whenever a value to be processed is available".

For instance, instead of thinking of several Observables each returning a single result for different XHR requests, think about one Observable containing a continuous stream of actionable XHR requests and their responses.  This is very similar to using event callbacks in the dom: you're really defining continuous behavior to continuous inputs.

**Comment by mmacfadden** [comment]
@corps Thanks for the input.  I am not confused about promises and observables.  I am fairly well versed in both, and understand the various operators.  In this instance I do not have a data flow.  I have a method whose semantics are that it always either A) returns a single value, or B) returns an error.  When I present an API like the below to a user:

```
 someMethod(): Promise<number>
```

They know that the semantics of the method are as such.  If however I return an observable:

```
 someMethod(): Observable<number>
```
It is no longer clear.  The API contract for this method states that the method could emit 0 values, 1 value or multiple values.  What's more it could emit any number of values or errors before being completed.  This is not in fact the semantics of the method, and returning an observable muddies the implied contract of the method.  The Promise, why less powerful actually communicates the correct semantics.  Doing things like take(1) or first() while logically work, from an API contract standpoint they are not ideal.

I believe this to be the reason that Single exists.  It gives the contract semantics of a promise but some of the power of the Observable.  I am using Single on the Java / Scala side and it works fairly well.  So my question still stands.

Has the RxJS community considered adding this concept?

**Comment by mmacfadden** [comment]
I would also add, that my premise was that I would love to ditch the promises.  I am just hesitant to do so because of the improper API semantics Observables have for things that are not actually data flows, but just single shot async request response type methods.

**Comment by corps** [comment]
If this is an API, then why not simply accept a callback itself?

```
someMethod(cb: (result, err)=>void)
```

This makes it most agnostic and makes the semantics extremely clear.  Users can still consume it with an Observable fairly trivially (bindCallback), but it's special result or error semantics are preserved.

**Comment by corps** [comment]
Actually, let me take that back, I think I understand better now what you're trying to say.

It seems like then.. you just want Promises?  Like my personal resistance to a Single like type is that it mudifies an already large proliferation of similar approaches to a problem.

The moment you want an object to quack like an Observable, but not behave like an Observable in one of its fundamental ways, makes me wonder if there isn't some emerging complexity that could be avoided by possibly just either using Promises instead or pushing the issue downstream with the callback approach?

**Comment by corps** [comment]
Also, single shot async request response type methods can definitely be modeled with a data stream.  It actually has rather elegant semantics, because you can, for instance, cancel duplicate requests without triggering an error.  Imagine multiple clicks to a button; with a Promise like semantic of 1 result or 1 error, you're forced to produce an error for each cancelled request, when all you're really concerned about is not producing an actionable response result for them.

**Comment by mmacfadden** [comment]
Thanks again for the thoughts!  I think the main reason we have been using promises are 1) there were all the rage when we started. 2) They are (limitedly) compoeable.  You can use then as a poor mans "map".  And there are a few operators like (all) out there, and 3) we had just been trying to fit in with a standard based approach.

With a callback you don't get a lot of the chaining benefits you get with promises or observables.

However, since then we have started to incorporate RxJS into the API, and I would prefer not to have two different concepts in there if possible.  Once we get into merging and mapping observables we find our self "fromPromise"ing our promises quite often.

Looking at the API if I had something like Single, I would probably just switch over to that immediately.  If not, we are still "weighing our options". No right or wrong answer.

**Comment by corps** [comment]
@mmacfadden One possible alternative, as I mentioned, is to actually leverage FRP here.

Instead of thinking of any of your async apis as "one offs" with strict single failure or success semantics, why not think of these apis as a single stream of responses given a stream of inputs?

```
function myWebApi(request$) {
  return Rx.Observable.create(function(observer) {
    var curRequest;
    return request$.subscribe(function(params) { 
      if(curRequest && !curRequest.completed) curRequest.abort();
      curRequest = ajax(url, params).then(observer.onNext, observer.onError);
    }, observer.onError, observer.onCompleted);
  });
}
```

This has a lot of nice benefits: you're no longer creating extraneous objects for each request, you have the ability to dedup requests that happen simultaneously, and you can feed request$ to test with any sort of Observable (fromArray with some delays is great here!)

Plus, it's just observables.

**Comment by mmacfadden** [comment]
@corps That's interesting an actually is pretty useful in a couple scenarios.  For the particular case, I am not doing any kind of ajax request.  I actually have a custom web socket protocol and a bunch of other internal stuff going on. It's not a traditional REST call at all.  The use case is a custom javascript client talking to a custom server side component over a websocket.  The end user of our API is not aware of the underlying semantics of the request and or response.  Part of the issue is that most of my background is in mostly in statically typed languages (at the moment Scala).  On the javascript side, I am using TypeScript primarily.  So I have things like interfaces, classes, and methods with return types, etc.

So for me, my main concern is not that I don't like what the Observable does.  I think its great.  Its that I don't like the semantics it suggests when I include it in certain parts of my API.  I have many places were I am using Observables to model data flows and it works well.  What I don't like is when I have a method called "getUser(userId)" and the semantics for the method is, it will always return exactly one value, or it will throw an error.  I also don't care about deduping requests, because once the data is gotten it is cached locally anyway, and this is not something where I am worried about DOM events or the user typing or something where I would need to debounce.

So I have those semantics in my brain.  When I do this:

```js
export class UserService {
  getUser(): Observable<User> { ... }
}
```

When a consumer reads this API they can not tell what the semantics of the method are. The possibilities include:

1. Emit 0 values, and complete.
1. Emit 1 value, and complete.
1. Emit N values and then complete.
1. Emit a value then an error, then complete.
1. Emit a error, and then complete.
1. etc. etc. etc.

With the Observable, the user can't look at the method and tell that the method has very specific semantics.  Coming from Scala... things like Try, Future, Promise, Stream, etc. all have similar functional composablility, but the have very specific semantics.  If my return type is a Future vs a Flow, I know the former will either succeed or fail, and the latter will behave like a stream.  What I am looking for is a way to better communicate the semantics of my methods via a return type that matches the semantics of the method.

I feel like an Observable, no matter what operators you apply to it, still does not meet that expectation for methods that which to describe a "return or error" contract.  I think this is specifically relevant to me because I am working with TypeScript.  If it were just JavaScript, I would only be worried about how the Observable "behaves", which is under my control with how I use it, since there is no return type.  However, in TyperScript there is a statically typed return type, and an API contract described at the language level.  And the contract of the Observable doesn't line up.

I do very much appreciate you taking the time to discuss this.  Again there is no wright or wrong answer.

**Comment by corps** [comment]
I understand better then.

If it's purpose is to communicate semantics, would it be possible to create a type Algebra in Typescript then without introducing a new specific implementation?

My main concern is just bloat of implementations.  For instance, if you wished the contract of one response or error to be enforced at the implementation, it creates kind of weird scenarios.  Does providing two values to a Single produce an error result /after/ producing a successful result?  Is a take(1) ever cast-able to a Single if the enforcement semantics won't actually be the same?

But if it's mostly type algebra to provide clarity, in truth I have no problem :P

**Comment by mmacfadden** [comment]
Yeah.  That might be an option.  I can look in to that.  I think that the folks in the Java/Scala side probably created Single, because that language didn't already have something like a Promise to begin with.
